File: solutions/853.most-profit-assigning-work/most-profit-assigning-work.cpp
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        const int n = difficulty.size();
        vector<pair<int, int>> jobs;
        for(int i = 0; i < n; i++) 
            jobs.emplace_back(difficulty[i], profit[i]);
        std :: sort(jobs.begin(), jobs.end());
        std :: sort(worker.begin(), worker.end());
        int ans = 0, best = 0, i = 0;
        for(int level : worker) {
            while(i < n && level >= jobs[i].first) 
                best = max(best, jobs[i++].second);
            ans += best;
        }
        return ans;
    }
};

File: solutions/853.most-profit-assigning-work/most-profit-assigning-work.java
class Solution {
    public int maxProfitAssignment(int[] difficulty, int[] profit, int[] worker) {
        int maxProfit[] = new int[100001];
        for(int i = 0; i < difficulty.length; i++) 
            maxProfit[difficulty[i]] = Math.max(maxProfit[difficulty[i]], profit[i]);
        for(int i = 2; i < 100001; i++) 
            maxProfit[i] = Math.max(maxProfit[i], maxProfit[i - 1]);
        int ans = 0;
        for(int level : worker) ans += maxProfit[level];
        return ans;
    }
}